src/gerbil/runtime/mop.ss

```diff
@@ -17,65 +17,121 @@ namespace: #f
+            precedence-list slot-vector slot-table properties constructor method))
+         (slot-vector
+          (list->vector (cons '##type slots)))
+         (slot-table
+          (let (slot-table (make-hash-table-eq))
+            (for-each
+              (lambda (slot field)
+                (hash-put! slot-table slot field)
+                (hash-put! slot-table (symbol->keyword slot) field))
+              slots
+              (iota (length slots) 1))
+            slot-table))
+         (flags
+          (##fxior type-flag-extensible type-flag-concrete type-flag-id
+                   class-type-flag-struct))
+         (fields ;; exclude those from ##type-type because this is for Gambit type use
+          ;; the field-flags of 5 mean "(1) printable, (4) equalable",
+          ;; the drop of 5 is the number of type-type slots
+          (list->vector (apply append (map (cut list <> 5 #f) (drop slots 5)))))
+         (properties
+          '((direct-slots: ,@slots) ;; include those from ##type-type because it's not a class
+            (struct: . #t)))
+         (t (##structure
+             #f            ; type: self reference, set below
+             class::t.id   ; type-id
+             'class        ; type-name
+             flags         ; type-flags
+             ##type-type   ; type-super
+             fields        ; type-fields
+             []            ; class-type-precedence-list
+             slot-vector   ; class-type-slot-vector
+             slot-table    ; class-type-slot-table
+             properties    ; class-type-properties
+             #f            ; class-type-constructor
+             #f)))         ; class-type-methods
+    (##structure-type-set! t t)         ; self reference
+    t))
+
+(def (class-type? obj)
+  (##structure-instance-of? obj class::t.id))
+
+(def (class-type-id klass)
   (cond
-   ((type-descriptor? klass) (##type-id klass))
+   ((class-type? klass) (##type-id klass))
    ((not klass) #f)
    (else (error "not a type descriptor" klass))))
 
-(def (type=? x y)
-  (eq? (type-id x) (type-id y)))
-
-(def (type-descriptor? klass)
-  (and (##type? klass)
-       (eq? (##structure-length klass) 12)))
-
-(def (type-struct? klass)
-  (##fxbit-set? 21 (##type-flags klass)))
-
-(def (type-final? klass)
-  (##fxbit-set? 22 (##type-flags klass)))
+(def (class-type=? x y)
+  (eq? (class-type-id x) (class-type-id y)))
 
-(def (class-type? klass)
-  (type-descriptor? klass))
+(def (type-opaque? type)
+  (##fxpositive? (##fxand (##type-flags type) type-flag-opaque)))
+(def (type-extensible? type)
+  (##fxpositive? (##fxand (##type-flags type) type-flag-extensible)))
+(def (class-type-final? type)
+  (##fxzero? (##fxand (##type-flags type) type-flag-extensible)))
+(def (class-type-struct? klass)
+  (##fxpositive? (##fxand (##type-flags klass) class-type-flag-struct)))
+(def (class-type-sealed? klass)
+  (##fxpositive? (##fxand (##type-flags klass) class-type-flag-sealed)))
```

Comment:
I am thinking `##fxbit-set?` might be faster but ok, let's go with this for now.